### ISSUE TITLE
Snippet, lineSnippet and multilineSnippet

### ISSUE THREAD

**Issue by michaelcfanning** [oasis-tcs/sarif-spec#197]
Our team has been attempting to implement the new snippet regions from v2. We've uncovered a subtlety in doing so. Generally, the snippet associated with a region is not very helpful (because this is typically a highlighted piece of code with the remainder of the source file surrounding it). In one example from yesterday, for example, using the Microsoft commercial C++ checker, this snippet was " = 0;"

This snippet, on its own in a browser viewer, is not very helpful. What we really want in our first experience is the single line of code in which this region appears. We don't want to embed a text snippet, however, that doesn't actually match the physical location region associated with the result. In addition to the single line snippet, in other cases, we need a multiline snippet, one that includes just a bit more code (for additional context, but also to provide a more stable partial fingerprint for result matching scenarios).

To make this work in our existing model, we'd effectively require three regions in a physical location:

region (precisely matches the reported location of the problem)
singleLineRegion (region matching a single line of code)
multilineRegion (multi-line region)

Some issues: This introduces a new property. Also, the singleLineRegion could be inferred from region in most cases (since the region will specify a start line). This approach is helpful for viewer consumption, however, since the spec will be very clear: for singleLine, emit the complete line of code that's most helpful in understanding the result. For multiline text snippet, emit a surrounding code region that best helps explain the problem.

Michael

**Comment by michaelcfanning** [comment]
@lgolding

**Comment by michaelcfanning** [comment]
So in our test implementation, we remove leading whitespace and trailing line break details. An interesting question is whether this kind of processing s/be done before emitting the snippet or when consuming it.

**Comment by lgolding** [comment]
@michaelcfanning 

To be clear, you're proposing to:
- Rename the existing `physicalLocation` property `contextRegion` to `multiLineRegion` for clarity, and
- Introduce `singleLineRegion`.

As you said: a viewer can easily expand `region` to the entire line, so I don't see what's gained by adding the new property `singleLineRegion`. Consider:
```
physicalLocation: {
  "region": {
    "startLine": 42,
    "startColumn": 25,
    "endColumn": 28
  },
  "contextRegion": {
    "startLine": 39,
    "endLine": 45
  }
}
```
Do we really gain anything by adding this?
```
physicalLocation: {
  "region": {
    "startLine": 42,
    "startColumn": 25,
    "endColumn": 28
  },
  "singleLineRegion": {
    "startLine": 42
  },
  "multiLineRegion": {
    "startLine": 39,
    "endLine": 45
  }
}
```
We _know_ that the single line is 42; `region` told us so. A viewer is welcome to highlight the line if it wants. The only time the greater explicitness helps is when the starting position is specified by `charOffset` rather than `startLine`/`startColumn`:
```
physicalLocation: {
  "region": {
    "charOffset": 312,
    "charLength": 3
  },
  "singleLineRegion": {
    "startLine": 42
  },
  "multiLineRegion": {
    "startLine": 39,
    "endLine": 45
  }
}
```
But again you don't need `singleLineRegion`, because it's legal to specify both `charOffset` and `startLine`:
```
physicalLocation: {
  "region": {
    "charOffset": 312,
    "charLength": 3,
    "startLine": 42
  },
  "multiLineRegion": {
    "startLine": 39,
    "endLine": 45
  }
}
```
Finally, I'm not wild about renaming `contextRegion` to `multiLineRegion`, because the existing `region` _is_ allowed to span multiple lines (although, as you've told me, that's not always a good visual experience). Renaming `contextRegion` to `multiLineRegion` suggests that `region` should _not_ have multiple lines.

I think the existing name is good, because it conveys what to me is most important about it: not that "this region has multiple lines" (a "physical" statement) but rather that "this region provides context around the result" (a "logical" statement).

**Comment by michaelcfanning** [comment]
The issue here is how to inject the source code into the log file. Yes, we know that line 42 is referenced by the region but we don't have the source code for it. We don't want to inject the whole file, that's what the snippet is for, it helps us avoid that.

Let me frame this another way: our testing shows that we have two distinct needs for snippets. In one, we show a single line of code. In another, we show a larger snippet, typically either 3 or 5 lines. If we don't take a change here, to support this scenario we'd be required to populate the context region with a multiline snippet and then to go grab whatever line within it matches the start line of the region.

that could work, but the region, of course, might not specify the start line, requiring more calculation.

**Comment by lgolding** [comment]
**TL;DR**: Ok, I see what you're saying. I think your suggested workaround is good enough but the TC should discuss.

**Details**: If the (existing) `region` property is:
```
"region": {
  "startLine": 15,
  "startColumn": 11,
  "endColumn": 14,
  "snippet": {
    "text": "= 0"
}
```
... and you want to highlight the entire line
```
    int k = 0;
```
... then how do you do it, if you don't have source code?

You propose one possible workaround: to use `contextRegion` (which would not _have_ to be multiline, but might be):
```
"contextRegion": {
  "startLine": 13,
  "endLine": 17,
  "snippet": {
    "text": "    int i;\r\n    int j;\r\n    int k = 0;\r\n    int m;\r\n    int n;\r\n"
  }
}
```
... and pick out the right line.

And as you say, that requires calculation if you have `charOffset` instead of `startLine`.

OTOH, if you know enough to put together a `contextRegion` like this, containing one or more full lines, then you probably know what line the issue is on. And `charOffset` is -- I believe -- less common than `startLine`/`startColumn`.

**Comment by michaelcfanning** [comment]
right, but a subtlety here is that you need to express startLine in both regions or you need to utilize charOffset in both. if you utilize startLine in one and charOffset in the other, you can't reliably locate the data. You could attempt to search for the original region snippet but it could be repeated in the broader region.

so if we stick with this as a pattern, you should author a nice validation rule along these lines. :)

**Comment by lgolding** [comment]
... oh. Sorry for stream of consciousness posts...

`contextRegion` is always good enough to allow you to find the _source code_ of the desired line, but it's not always enough to find the _line number_ of the desired line (as in your `charOffset`/`startLine` mismatch scenario).

**Comment by michaelcfanning** [comment]
No, you can't even assume you can locate the source code. Consider trying to retrieve the first or last line of a file. You might reasonably want to take a following line or two (if you're at the beginning of the file) or the preceding line or two (if you're at the end). You can't easily distinguish these cases. Consider the following source file:

"a\r\na\r\na\r\n"

And now imagine you want to fire against the first "a" in the file. This is contrived, I know. But it makes the point that it's best for the region to be minimally populated (and consist between the region and contextRegion) to support locating the complete line associated with the region.
